core/src/main/java/io/questdb/cutlass/pgwire/CircuitBreakerRegistry.java

```diff
@@ -0,0 +1,127 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+package io.questdb.cutlass.pgwire;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.CairoException;
+import io.questdb.cairo.sql.NetworkSqlExecutionCircuitBreaker;
+import io.questdb.mp.SimpleSpinLock;
+import io.questdb.std.*;
+
+import java.io.Closeable;
+
+// registry of circuit breakers utilized by pg wire contexts 
+// makes it possible to cancel queries executed in other connections/threads
+public class CircuitBreakerRegistry implements Closeable {
+
+    private final ObjList<NetworkSqlExecutionCircuitBreaker> circuitBreakers;
+    private final IntList freeIdx;
+
+    private final SimpleSpinLock lock;
+
+    private final Rnd random;
+
+    public CircuitBreakerRegistry(PGWireConfiguration configuration, CairoConfiguration cairoConfig) {
+        lock = new SimpleSpinLock();
+        int limit = configuration.getDispatcherConfiguration().getLimit();
+        circuitBreakers = new ObjList<>(limit);
+        freeIdx = new IntList(limit);
+        if (configuration.getRandom() != null) {
+            // use separate rnd to avoid changing existing pg test results 
+            random = new Rnd(configuration.getRandom().getSeed0(), configuration.getRandom().getSeed1());
+        } else {
+            random = new Rnd(
+                    cairoConfig.getNanosecondClock().getTicks(),
+                    cairoConfig.getMicrosecondClock().getTicks());
+        }
+
+        for (int i = 0; i < limit; i++) {
+            circuitBreakers.add(new NetworkSqlExecutionCircuitBreaker(configuration.getCircuitBreakerConfiguration(), MemoryTag.NATIVE_CB5));
+            freeIdx.add(i);
+        }
+    }
+
+    public int acquire() {
+        int idx;
+        int secret;
+        lock.lock();
+        try {
+            int size = freeIdx.size();
```

Comment:
cannot a race between:
1. opening a new pg connection
2. closing an old connection
3. releasing a circuit breaker
cause this line to return 0 and explode one line later?

Reply:
Good question! :)

Reply:
I might be wrong but don't think that's possible because IODispatcher.accept is guarded runSerially() and context close happens before decrementing number of active connections .